### PR TITLE
*: fix wrong flen of  table information_schema.placement_rules

### LINKED ISSUE #29482: infoschema.placement_rules table uses wrong flen values
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
tidb> show create table placement_rules\G
*************************** 1. row ***************************
       Table: PLACEMENT_RULES
Create Table: CREATE TABLE `PLACEMENT_RULES` (
  `POLICY_ID` bigint(64) NOT NULL,
  `CATALOG_NAME` varchar(512) NOT NULL,
  `POLICY_NAME` varchar(5) DEFAULT NULL,
  `SCHEMA_NAME` varchar(5) DEFAULT NULL,
  `TABLE_NAME` varchar(5) DEFAULT NULL,
  `PARTITION_NAME` varchar(5) DEFAULT NULL,
  `PRIMARY_REGION` varchar(5) NOT NULL,
  `REGIONS` varchar(5) NOT NULL,
  `CONSTRAINTS` varchar(5) NOT NULL,
  `LEADER_CONSTRAINTS` varchar(5) NOT NULL,
  `FOLLOWER_CONSTRAINTS` varchar(5) NOT NULL,
  `LEARNER_CONSTRAINTS` varchar(5) NOT NULL,
  `SCHEDULE` varchar(20) NOT NULL,
  `FOLLOWERS` bigint(64) NOT NULL,
  `LEARNERS` bigint(64) NOT NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
1 row in set (0.00 sec)
```

### 2. What did you expect to see? (Required)

The policy_name, schema name, table_name and partition name for example should probably be varchar(64). The others might need to be re-evaluated (primary region, regions).

### 3. What did you see instead (Required)

VARCHAR(5) etc.

### 4. What is your TiDB version? (Required)

```sql
tidb> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v5.3.0-alpha-1298-gec7a638e3-dirty
Edition: Community
Git Commit Hash: ec7a638e33fdde89d07f50b7ad587bc00d1d94f0
Git Branch: master
UTC Build Time: 2021-11-02 02:19:11
GoVersion: go1.16.9
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.00 sec)

```

### PR DESCRIPTION
**Pull request by sylzd**
### What problem does this PR solve?

Issue Number: close #29482

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: infoschema/tables.go [hunk 1]
```diff
@@ -387,7 +387,7 @@ var tablesCols = []columnInfo{
 	{name: "TIDB_ROW_ID_SHARDING_INFO", tp: mysql.TypeVarchar, size: 255},
 	{name: "TIDB_PK_TYPE", tp: mysql.TypeVarchar, size: 64},
 	{name: "TIDB_PLACEMENT_POLICY_NAME", tp: mysql.TypeVarchar, size: 64},
-	{name: "TIDB_DIRECT_PLACEMENT", tp: mysql.TypeVarchar, size: types.UnspecifiedLength},
+	{name: "TIDB_DIRECT_PLACEMENT", tp: mysql.TypeVarchar, size: 512},
```
Comment (mjonss) [inline infoschema/tables.go:390]: Is 512 the maximum length of Direct Placement? I'm not sure there even is a maximum?

File: infoschema/tables.go [hunk 2]
```diff
@@ -560,7 +560,7 @@ var partitionsCols = []columnInfo{
 	{name: "TABLESPACE_NAME", tp: mysql.TypeVarchar, size: 64},
 	{name: "TIDB_PARTITION_ID", tp: mysql.TypeLonglong, size: 21},
 	{name: "TIDB_PLACEMENT_POLICY_NAME", tp: mysql.TypeVarchar, size: 64},
-	{name: "TIDB_DIRECT_PLACEMENT", tp: mysql.TypeVarchar, size: types.UnspecifiedLength},
+	{name: "TIDB_DIRECT_PLACEMENT", tp: mysql.TypeVarchar, size: 512},
```
Comment (mjonss) [inline infoschema/tables.go:563]: Same here, I'm not sure if there actually is a maximum for Direct Placement, even though 512 would cover most cases, there might be more complex ones that would not be shown.

File: infoschema/tables.go [hunk 3]
```diff
@@ -1451,16 +1451,16 @@ var tableAttributesCols = []columnInfo{
...
-	{name: "REGIONS", tp: mysql.TypeVarchar, size: types.UnspecifiedLength, flag: mysql.NotNullFlag},
-	{name: "CONSTRAINTS", tp: mysql.TypeVarchar, size: types.UnspecifiedLength, flag: mysql.NotNullFlag},
-	{name: "LEADER_CONSTRAINTS", tp: mysql.TypeVarchar, size: types.UnspecifiedLength, flag: mysql.NotNullFlag},
-	{name: "FOLLOWER_CONSTRAINTS", tp: mysql.TypeVarchar, size: types.UnspecifiedLength, flag: mysql.NotNullFlag},
-	{name: "LEARNER_CONSTRAINTS", tp: mysql.TypeVarchar, size: types.UnspecifiedLength, flag: mysql.NotNullFlag},
+	{name: "POLICY_NAME", tp: mysql.TypeVarchar, size: 64},    // Catalog wide policy
+	{name: "SCHEMA_NAME", tp: mysql.TypeVarchar, size: 64},    // System policy does not have a schema
+	{name: "TABLE_NAME", tp: mysql.TypeVarchar, size: 64},     // Schema level rules does not have a table
+	{name: "PARTITION_NAME", tp: mysql.TypeVarchar, size: 64}, // Table level rules does not have a partition
+	{name: "PRIMARY_REGION", tp: mysql.TypeVarchar, size: 64, flag: mysql.NotNullFlag},
```
Comment (mjonss) [inline infoschema/tables.go:1458]: Do we have some design/documentation on the maximum lengths?

File: infoschema/tables.go [hunk 4]
```diff
@@ -1451,16 +1451,16 @@ var tableAttributesCols = []columnInfo{
...
+	{name: "REGIONS", tp: mysql.TypeVarchar, size: 64, flag: mysql.NotNullFlag},
```
Comment (mjonss) [inline infoschema/tables.go:1459]: I think this is too small.

File: infoschema/tables.go [hunk 5]
```diff
@@ -1451,16 +1451,16 @@ var tableAttributesCols = []columnInfo{
...
+	{name: "CONSTRAINTS", tp: mysql.TypeVarchar, size: 64, flag: mysql.NotNullFlag},
+	{name: "LEADER_CONSTRAINTS", tp: mysql.TypeVarchar, size: 64, flag: mysql.NotNullFlag},
+	{name: "FOLLOWER_CONSTRAINTS", tp: mysql.TypeVarchar, size: 64, flag: mysql.NotNullFlag},
+	{name: "LEARNER_CONSTRAINTS", tp: mysql.TypeVarchar, size: 64, flag: mysql.NotNullFlag},
```
Comment (mjonss) [inline infoschema/tables.go:1463]: Constraints can be quite long. I don't think we are blocking arbitrary lengths of constraints? Or does there exist a limit in PD or TiDB?

**Review by mjonss — COMMENTED** [review]
For the identifier lengths I agree (and sorry for not setting them in the first place).
But for the rules them selfs I'm not sure 64/512 is the right values. Maybe a `text` type would be better? @morgo what do you think?

File: infoschema/tables.go (hunk 1 above)
Comment (sylzd) [reply infoschema/tables.go:390]: It deserves discussing, IMO, the maximum of varchar 16383 is enough, and this table is small, we do not need care the storage overhead too much.

File: infoschema/tables.go (hunk 2 above)
Comment (sylzd) [reply infoschema/tables.go:563]: ditto.

File: infoschema/tables.go (hunk 3 above)
Comment (sylzd) [reply infoschema/tables.go:1458]: @morgo

File: infoschema/tables.go (hunk 4 above)
Comment (sylzd) [reply infoschema/tables.go:1459]: can we use 16383 for all items that may have a big size?

File: infoschema/tables.go (hunk 1 above)
Comment (morgo) [reply infoschema/tables.go:390]: I like 512 as a max. We can always increase it later (I might have also suggested 1024), but decreasing it is difficult.

File: infoschema/tables.go (hunk 2 above)
Comment (morgo) [reply infoschema/tables.go:563]: Same here: 512.

File: infoschema/tables.go (hunk 3 above)
Comment (morgo) [reply infoschema/tables.go:1458]: We can add it to the design document as 512. For identifiers, it is always 64, but there are no declared rules for length for other objects currently.

File: infoschema/tables.go (hunk 4 above)
Comment (morgo) [reply infoschema/tables.go:1459]: I prefer 512 (or 1024) here too.

File: infoschema/tables.go (hunk 5 above)
Comment (morgo) [reply infoschema/tables.go:1463]: There is no specific problem with allowing large values that I'm aware of. But my rationale for 512/1024 was that this is the limit for a key prefix on a text/blob column in MySQL, and in some cases we might want to index constraints.

(OK there is no such direct relationship: but the limit in MySQL feels about right, and does not seem to pose as a major limitation).

Comment (sylzd) [reply infoschema/tables.go:1463]: I agree that~

**Review by morgo — APPROVED** [review]

**Review by AilinKid — APPROVED** [review]

**Review by mjonss — APPROVED** [review]
LGTM

**Pull request merged**